File: packages/client/src/ffi/directories.ts
import { toast } from "../utils";
import { invoke } from "./invoke";

export type Directory = {
  id?: number;
  uid: string;
  name: string;
  path: string;
  workspace_uid: string;
  parent_uid?: string;
  created_at: string;
  updated_at: string;
  synced?: boolean;
};

type Args = {
  list_dirs: { workspaceUid: string; parentUid?: string };
  get_directory_id: { workspaceUid: string; path: string };
  create_dir: {
    name: string;
    workspaceUid: string;
    path: string;
    parentUid?: string;
    uid?: string;
    created_at?: string;
    updated_at?: string;
    synced?: boolean;
  };
  delete_dir: { uid: string };
  update_dir: {
    uid: string;
    name: string;
    workspaceUid: string;
    path: string;
    parentUid?: string;
    synced?: boolean;
  };
};

export const createDir = async (
  name: string,
  workspaceUid: string,
  path: string,
  parentUid?: string,
  uid?: string,
  createdAt?: string,
  updatedAt?: string,
  synced?: boolean,
) => {
  try {
    const directory = await invoke<Args, Directory | null>()("create_dir", {
      name,
      path,
      workspaceUid,
      parentUid,
      uid,
      created_at: createdAt,
      updated_at: updatedAt,
      synced,
    });

    if (!directory) {
      toast("Failed to create directory!");
      return;
    }

    return directory;
  } catch (e) {
    console.error("[Directory] Create failed!", e);
    toast("Failed to create directory!");
  }
};

export const getDirs = async (workspaceUid: string, parentUid?: string) => {
  try {
    const directories = await invoke<Args, Array<Directory>>()("list_dirs", {
      workspaceUid,
      parentUid,
    });

    if (!directories) {
      throw "Unknown failure!";
    }

    return directories;
  } catch (e) {
    console.error("[Directory] Failed to fetch!", e);
  }
};

export const deleteDir = async (uid: string) => {
  try {
    if (
      !(await invoke<Args, boolean>()("delete_dir", {
        uid,
      }))
    ) {
      toast("Failed to delete directory!");
    }

    return;
  } catch (e) {
    console.error("[Directory] Failed to delete!", e);
    toast("Failed to delete directory!");
  }
};

export const getByPath = async (path: string, workspaceUid: string) => {
  try {
    return await invoke<Args, number | undefined>()("get_directory_id", {
      workspaceUid,
      path,
    });
  } catch (e) {
    console.error("[Directory] Failed to fetch by path!", e);
  }
};

export const updateDir = async (
  uid: string,
  name: string,
  workspaceUid: string,
  path: string,
  parentUid?: string,
  synced?: boolean,
) => {
  try {
    const directory = await invoke<Args, Directory | null>()("update_dir", {
      uid,
      name,
      workspaceUid,
      path,
      parentUid,
      synced,
    });

    if (!directory) {
      toast("Failed to update directory!");
      return;
    }

    return directory;
  } catch (e) {
    console.error("[Directory] Failed to update!", e);
    toast("Failed to update directory!");
  }
};

File: crates/pyxis-app/src/handlers/directories.rs
use chrono::Utc;
use pyxis_db::{database::Database, entities::directories::Directory};
use tauri::State;

#[tauri::command]
pub fn create_dir(
    name: String,
    workspace_uid: String,
    path: String,
    parent_uid: Option<String>,
    created_at: Option<String>,
    updated_at: Option<String>,
    uid: Option<String>,
    database: State<Database>,
) -> Option<Directory> {
    let directory = Directory::new(name, workspace_uid, path, parent_uid, None, created_at, updated_at, uid);

    match directory.create(&database.get_connection()) {
        Ok(_) => Some(directory),
        Err(e) => {
            eprintln!("[Directories] Failed to create! {}", e);
            None
        }
    }
}

#[tauri::command]
pub fn list_dirs(
    workspace_uid: String,
    parent_uid: Option<String>,
    database: State<Database>,
) -> Option<Vec<Directory>> {
    match Directory::list(&database.get_connection(), workspace_uid, parent_uid) {
        Ok(directories) => Some(directories),
        Err(e) => {
            eprintln!("[Directories] Failed to fetch! Error: {e}");
            None
        }
    }
}

#[tauri::command]
pub fn delete_dir(uid: String, database: State<Database>) -> bool {
    match Directory::delete(uid, &database.get_connection()) {
        Ok(_) => true,
        Err(e) => {
            eprintln!("[Directories] Failed to delete! {}", e);
            false
        }
    }
}

#[tauri::command]
pub fn update_dir(
    id: i32,
    name: String,
    workspace_uid: String,
    path: String,
    parent_uid: Option<String>,
    database: State<Database>,
) -> Option<Directory> {
    let conn = &database.get_connection();
    let directory = match Directory::list(conn, workspace_uid.clone(), parent_uid.clone()) {
        Ok(w) => w.into_iter().find(|w| w.id == Some(id)),
        Err(e) => {
            eprintln!("[Directories] Failed to get for update! {}", e);
            None
        }
    };

    if let Some(mut dir) = directory {
        dir.name = name;
        dir.workspace_uid = workspace_uid;
        dir.path = path;
        dir.parent_uid = parent_uid;
        dir.updated_at = Utc::now().to_rfc3339();

        match dir.update(conn) {
            Ok(_) => {
                return Some(dir);
            }
            Err(e) => {
                eprintln!("[Directories] Failed to update! {}", e);
                return None;
            }
        }
    }

    None
}

#[tauri::command]
pub fn get_directory_id(
    path: String,
    workspace_uid: String,
    database: State<Database>,
) -> Option<i64> {
    match Directory::get_by_path(&database.get_connection(), path, workspace_uid) {
        Ok(directory_id) => Some(directory_id),
        Err(e) => {
            eprintln!("[Directory] Failed to fetch id! Error: {e}");
            None
        }
    }
}
